Ranking solutions.

marcovc opened the pull request:
* Collects a vector of estimated prices from a PriceEstimating.
* Sends vector to objective_value()
* Changes base for computing objective from U256 -> BigRational

### Test Plan
Unit test.

marcovc committed:
Draft: Ranking solutions.

marcovc committed:
Not done yet comment.

marcovc committed:
Completes objective value v2.

marcovc committed:
Change base to BigRational for computing objective.

marcovc committed:
Fix U256 -> BigRational in tests.

marcovc committed:
Merge branch 'main' into compare-solutions

marcovc committed:
WIP

marcovc committed:
Merge branch 'main' into compare-solutions

marcovc committed:
Connects price estimation with objective computation.

Commit:
Merge branch 'main' into compare-solutions

marcovc committed:
Handles case where there are no limit orders.

fleupold commented on solver/src/driver.rs:
```diff
@@ -63,6 +68,35 @@ impl Driver {
+
+        // For ranking purposes it doesn't matter how the external price vector is scaled.
+        // If we use native_token here instead, should it come as another ctor parameter?
+        let denominator_token: H160 = tokens[0];
```
I think it makes sense to use the native token here, as it will likely be the most liquid and therefore give the most accurate price estimations. Can be passed into the driver as ctor arg as you suggested.

marcovc replied:
Yes, but notice that we don't care how accurate the actual price estimations are for the purposes of ranking. On the other hand, it might be better for logging/debugging. Done here e41c35.

fleupold commented on solver/src/settlement.rs:
```diff
@@ -205,98 +258,117 @@ mod tests {
         // Two goods are worth the same (100 each). If we were willing to pay up to 60 to receive 50,
         // but ended paying the price (1) we have a surplus of 10 sell units, so a total surplus of 1000.
+
+        // I hope there is a better way :)
```
You could define yourself a macro or a small helper function.

marcovc replied:
Done in a2a4e.

fleupold commented on solver/src/settlement.rs:
```diff
@@ -139,35 +180,47 @@ impl Settlement {
         .checked_mul(sell_token_price)?
-        .checked_sub(executed_amount.checked_mul(buy_token_price)?)
+        .checked_sub(&executed_amount.checked_mul(buy_token_price)?)?;
+    // Should we simply return 0 when the order fails to satisfy the limit price,
```
I think it's fine to treat this the same way as an overflow in the computation. We should probably log an error so we are being made aware of it.

marcovc replied:
Done!

marcovc replied:
Done here: 119b062

fleupold commented on solver/src/settlement.rs:
```diff
@@ -111,26 +117,61 @@ impl Settlement {
+    }
+
+    // Objective is scaled by (harmonic) mean of external/found price ratio.
+    #[allow(dead_code)]
```
I think we should check this in on a separate branch/commit but not have dead_code in the main codebase.

marcovc replied:
Removed v2 here: f97dd2c.

fleupold reviewed (commented):
Looks really good! I don't think extra end to end tests are necessary. Maybe you can use the simple examples from https://docs.google.com/document/d/1Toc7DGA8z3pfSBeVjeF9xl5VsWRarx-8K2O5O_aSSDY/edit and implement them as a test case to make sure we have the relative ordering between two different solutions right?

vkgnosis commented on solver/src/settlement.rs:
```diff
@@ -205,98 +258,117 @@ mod tests {
         // Two goods are worth the same (100 each). If we were willing to pay up to 60 to receive 50,
         // but ended paying the price (1) we have a surplus of 10 sell units, so a total surplus of 1000.
+
+        // I hope there is a better way :)
+        let _5000 = BigRational::from_u32(5000).unwrap();
```
I don't think so because BigRational does not implement From<u32> and so on directly. I would use a small helper functions (or macro) here to avoid the repetition. Something like 

```rust
        fn r(u: u128) -> BigRational {
            BigRational::from_u128(u).unwrap()
        }

        let _5000 = r(5000);
```

and at that point you can put that directly in the assert instead of going through a variable.

marcovc replied:
Thanks! It's here a2a4e.

marcovc commented on solver/src/settlement.rs:
```diff
@@ -111,26 +117,61 @@ impl Settlement {
-                .expect("Solution with trade but without price for sell token");
-            acc?.checked_add(trade.surplus(*sell_token_price, *buy_token_price)?)
-        }) {
+                .expect("Solution with trade but without price for buy token");
+            acc?.checked_add(&trade.surplus(sell_token_external_price, buy_token_external_price)?)
```
I thought about that too (btw I haven't found an easy to check that from the code since Ratio<T> is coded for a generic T which doesn't overflow for the specific case of T=BigInt).

However. I believe divisions by 0 are still coerced into None, so we would still need the Option...

marcovc replied:
Actually I'm not sure about what happens for div 0. Will check the source!

marcovc replied:
Found it: https://docs.rs/num-rational/0.1.42/num_rational/struct.Ratio.html#method.checked_div . I guess this means we need to keep the Option, right?

vkgnosis replied:
I still see a bunch of checked_... that are not checked_div so we could start by removing those and seeing how many options we can remove. For the divs it depends on whether the divisor can really be 0 which I am not sure about from the top of my head.

marcovc replied:
Done here:  119b062

marcovc committed:
Scales external price vector by native token.

marcovc committed:
Better solution to literal BigRational's.

marcovc committed:
Removed objective_v2.

Will be added later to a bookeeping branch.

marcovc committed:
Sort by cached key optimization.

marcovc committed:
Fixes score function and adds a unit test.

marcovc commented on solver/src/driver.rs:
```diff
@@ -63,6 +68,35 @@ impl Driver {
+            return Ok(hashmap![]);
+        }
+        // Computes set of traded tokens (limit orders only).
+        let tokens: Vec<H160> = limit_orders
+            .iter()
+            .flat_map(|lo| vec![lo.sell_token, lo.buy_token].into_iter())
```
Perhaps this is one of those cases where the penalty in efficiency doesn't compensate the penalty in readability?

vkgnosis replied:
I don't have a strong preference, just slightly prefer the non allocating version. So if you prefer this readability then it's fine if we keep it. Maybe compiler optimizes this out anyway.

marcovc replied:
What do you think about either implementing a small function or using something like this: https://danielkeep.github.io/rust-grabbag/doc/grabbag_macros/macro.collect!.html ?

bh2smith replied:
Is this collect more robust that the standard collect?

vkgnosis replied:
How does that macro help here?

marcovc replied:
How about this? b83fc75

vkgnosis replied:
I see. I'm fine with although it will hopefully no longer be needed in the next rust version.

marcovc commented on solver/src/driver.rs:
```diff
@@ -97,16 +134,15 @@ impl Driver {
-            Reverse(settlement.objective_value(/* estimated_prices*/))
-        });
+        settlements
+            .sort_by_key(|(_, settlement)| Reverse(settlement.objective_value(&estimated_prices)));
```
Done in fd1b65. Perhaps an even better solution would be to store the objective value in the settlement object upon construction, since otherwise we still compute it twice .

marcovc commented:
Fixed the main function that scores the settlement - we can't simply recompute using external price vectors, but instead we need to use internal prices for computing surplus, and only in the end rescale using external prices. An extreme example where this makes a difference is when the external price vector fails to satisfy limit constraints.

Added a unit test.

marcovc committed:
Linter.

marcovc commented:
@twalth3r It would be great if you could also check this PR. Thanks!

bh2smith commented on solver/src/settlement.rs:
```diff
@@ -204,99 +251,249 @@ mod tests {
     }
 
+    // Helper function to save some repeatition below.
+    fn r(u: u128) -> BigRational {
```
Nit: I guess this has been reduced to just r because calling it rat would also lead the reader to have to check what it is anyway and rationalize is far too long. If not, I might suggest calling it rational or at the very least rat. I know it's mathematically sound and much more concise as is, but I'm just hoping to prevent the reader from actually having to jump to the definition.

marcovc replied:
I think calling it "rat" could tranform

https://github.com/gnosis/gp-v2-services/blob/79b12404652c7467002858ff079aeb0b5d8924dd/solver/src/settlement.rs#L406-L447

into

![tenor](https://user-images.githubusercontent.com/624308/111584736-2933f280-87b6-11eb-8119-8aaf7eef0a9d.gif)

:)

(now seriously, @vkgnosis suggested the name, but I think one letter is already too many letters - there should be an implicit conversion in this case)

vkgnosis replied:
> hoping to prevent the reader from actually having to jump to the definition

The function is one line long and just above the tests so my personal preference is as short as possible because it makes the asserts more readable.

bh2smith commented on solver/src/driver.rs:
```diff
@@ -9,15 +9,19 @@ use anyhow::{Context, Result};
 use futures::future::join_all;
 use gas_estimation::GasPriceEstimating;
+use itertools::Itertools;
+use maplit::hashmap;
```
What the difference between this and the standard library hash map?

marcovc replied:
I'm guessing that the HashMap is the struct, and this maplit::hashmap is just a macro to create literal hashmaps?

vkgnosis replied:
That's correct and in these cases I would suggest just checking the docs. From the use we see that the crate is called `maplit` so searching for that on crates.io or docs.rs shows the docs where it is explained.

bh2smith commented on solver/src/settlement.rs:
```diff
@@ -111,26 +117,53 @@ impl Settlement {
+    }
+
+    // For now this computes the total surplus of all EOA trades.
+    pub fn objective_value(&self, external_prices: &HashMap<H160, BigRational>) -> BigRational {
```
I see that this method is now just encapsulating its recomputed counterpart with an additional overflow handling, do we really need this or could it all fit into a single place?

marcovc replied:
Yes, fixed in 779d70 . Thanks!

bh2smith reviewed (approved):
Really nice work here! Made a few comments and questions that could be treated as suggestions but no strong opinions either way.

twalth3r reviewed (approved):
Looks all good to me, thanks!

marcovc committed:
Removes redundant method.

marcovc committed:
Removed redundant checked arith.

marcovc committed:
Better comment.

marcovc committed:
Playing with macros.

marcovc committed:
Fixed macro.

marcovc committed:
Linter.

marcovc committed:
Improves macro and moves it to a "util.rs" file.

fleupold commented on solver/src/driver.rs:
```diff
@@ -63,6 +71,36 @@ impl Driver {
         }
     }
 
+    // Should this go here or on price_estimate.rs?
```
no, I think this is very specific to the driver, and should stay here.

fleupold commented on solver/src/driver.rs:
```diff
@@ -80,11 +118,14 @@ impl Driver {
         tracing::debug!("got {} AMMs", amms.len());
 
         let liquidity: Vec<Liquidity> = limit_orders
+            .clone()
```
nanonit: if you move `self.collect_estimated_prices(&limit_orders).await?;` above you shouldn't need to clone.

fleupold commented on solver/src/util.rs:
```diff
@@ -0,0 +1,22 @@
+// This macro allows to build literal iterations without allocating
+// a container such as a vector:
+// chain![1, 2, 3]
```
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia1.giphy.com%2Fmedia%2F4Z5lEBKx7WMfEja9In%2Fgiphy.gif&token=chapeau" data-gitmeme-token="chapeau"><img src="https://media1.giphy.com/media/4Z5lEBKx7WMfEja9In/giphy.gif" title="Created by gitme.me with /chapeau"/></a>

fleupold commented on solver/src/settlement.rs:
```diff
@@ -36,21 +38,25 @@ impl Trade {
+        &self,
+        sell_token_price: &BigRational,
+        buy_token_price: &BigRational,
+    ) -> Option<BigRational> {
```
nit: I wonder if this should return a Result instead of an Option so that we can pass the information about which trade is causing problems which would make it easier to debug. Can happen in a separate PR though as it's already an Option in the current codebase.

marcovc replied:
Will do in subsequent PR.

fleupold reviewed (approved):
Really good stuff!

marcovc committed:
Removing comment and small optimization.

Commit:
Merge branch 'main' into compare-solutions
